dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+	},
+	/* failure states */
+	[RFS_FAILED] = {
+		.sd_name      = "RFS_FAILED",
+		.sd_allowed   = M0_BITS(RFS_DONE),
+		.sd_flags     = M0_SDF_FAILURE,
+	},
+
+	/* intermediate states */
+#define _ST(name, allowed)            \
+	[name] = {                    \
+		.sd_name    = #name,  \
+		.sd_allowed = allowed \
+	}
+	_ST(RFS_WAITING,           M0_BITS(RFS_DONE,
+					   RFS_FAILED, RFS_WAITING)),
+#undef _ST
+};
+
+const static struct m0_sm_conf recovery_fom_conf = {
+	.scf_name      = "recovery_fom",
+	.scf_nr_states = ARRAY_SIZE(recovery_fom_states),
+	.scf_state     = recovery_fom_states,
+};
+
+static size_t recovery_fom_locality(const struct m0_fom *fom)
+{
+	return 1;
+}
+
+static const struct m0_fom_ops recovery_fom_ops = {
+	.fo_fini          = recovery_fom_self_fini,
+	.fo_tick          = recovery_fom_tick,
+	.fo_home_locality = recovery_fom_locality
+};
+
+static struct m0_fom_type recovery_fom_type;
+static const struct m0_fom_type_ops recovery_fom_type_ops = {};
+
+M0_INTERNAL int m0_drm_domain_init(void)
+{
+	int         rc = 0;
+
+	if (!m0_sm_conf_is_initialized(&m0_drm_sm_conf)) {
+		m0_fom_type_init(&recovery_fom_type,
+				 M0_DTM0_RECOVERY_FOM_OPCODE,
+				 &recovery_fom_type_ops,
+				 &dtm0_service_type,
+				 &recovery_fom_conf);
+
+		m0_sm_conf_init(&m0_drm_sm_conf);
+		rc = m0_sm_addb2_init(&m0_drm_sm_conf,
+				      M0_AVI_DRM_SM_STATE,
+				      M0_AVI_DRM_SM_COUNTER);
+
+		M0_POST(m0_sm_conf_is_initialized(&m0_drm_sm_conf));
+	}
+
+	return M0_RC(rc);
+}
+
+M0_INTERNAL void m0_drm_domain_fini(void)
+{
+	if (m0_sm_conf_is_initialized(&m0_drm_sm_conf)) {
+		m0_sm_addb2_fini(&m0_drm_sm_conf);
+		m0_sm_conf_fini(&m0_drm_sm_conf);
+		M0_POST(!m0_sm_conf_is_initialized(&m0_drm_sm_conf));
+	}
+}
+
+static void eolq_post(struct m0_dtm0_recovery_machine *m,
+		      struct eolq_item                *item)
+{
+	struct recovery_fom *rf = recovery_fom_local(m);
+	m0_be_queue_lock(&rf->rf_eolq);
+	/*
+	 * Do not post any events if we have already recovered this DTM0
+	 * service (eoq is an indicator here).
+	 */
+	if (!rf->rf_eolq.bq_the_end) {
+		/*
+		 * Assumption: the queue never gets full.
+		 * XXX: We could panic in this case. Right now, the HA/REDO FOM
+		 * should get stuck in the tick. Prorably, panic is
+		 * a better solution here.
+		 */
+		M0_BE_OP_SYNC(op, M0_BE_QUEUE_PUT(&rf->rf_eolq, &op, item));
+	}
+	m0_be_queue_unlock(&rf->rf_eolq);
+}
+
+static void heq_post(struct recovery_fom *rf, enum m0_ha_obj_state state)
+{
+	uint64_t event = state;
+	m0_be_queue_lock(&rf->rf_heq);
+	/*
+	 * Assumption: the queue never gets full.
+	 * XXX: We could panic in this case. Right now, the HA FOM should get
+	 *      stuck in the tick. Prorably, panic is a better solution
+	 *      here.
+	 */
+	M0_BE_OP_SYNC(op, M0_BE_QUEUE_PUT(&rf->rf_heq, &op, &event));
+	m0_be_queue_unlock(&rf->rf_heq);
+
+	M0_LOG(M0_DEBUG, "heq_enq " FID_F " %s ",
+	       FID_P(&rf->rf_tgt_svc),
+	       m0_ha_state2str(state));
+
+	if (!rf->rf_is_local)
```

Comment:
I don't understand this.
Maybe some comments will help me to understand:
Why is this event replicated and sent to local recovery fom too?

Reply:
3:working on it -- adding comment in the code.

Reply:
Fixed/merged.